### PR TITLE
Expose connection hot zones in `GraphNode`

### LINKED ISSUE #3009: Accessing Port from GraphNode
### Describe the project you are working on

[Finite State Machine](https://github.com/mechPenSketch/FiniteStateMachine), specifically in this context its main screen editor.

### Describe the problem or limitation you are having in your project

I can't drag the port of the GraphNode except by clicking directly on it. My intention is for the user to be able to port drag simply by clicking and dragging from either edge of the GraphNode.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add new methods in the GraphEdit that redifines what a hot zone is to click and drag a connection port.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

On the GraphEdit's cpp file, add two methods:

For input:
```
bool GraphEdit::is_in_input_hotzone(GraphNode gn, int i){
    Vector2 pos = gn->get_connection_input_position(j) + gn->get_position();
    return is_in_port_hotzone(pos / zoom, p_point / zoom);
}
```

and output
```
bool GraphEdit::is_in_output_hotzone(GraphNode gn, int i){
    Vector2 pos = gn->get_connection_output_position(j) + gn->get_position();
    return is_in_port_hotzone(pos / zoom, p_point / zoom);
}
```

Where `is_in_port_hotzone(...)` is renamed from `is_in_hot_zone(...)` to specify that the hot zone is related to the port by position.

`is_in_input_hotzone(...)` and `is_in_output_hotzone(...)` can then be exposed to be overridden if users want to redefine where the hotzone is, whether it would involve a port's position or not. 

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I can think of a few:
1. Adjusting the port offset as well as setting the port texture as a large transparent block. However, I will need to get the texture height right or there will be an awkward situation where by mousing over the edge where the texture doesn't overlap with the area of the GraphNode, the hover effect could appear, but clicking won't work. I would also have to re-adjust the texture's height should more than one line be used.
2. Restarting the main screen FSM from scratch where instead of using GraphEdit and GraphNode, I use a combination of PanelContainer, LineEdit and Line2D. Please don't make me restart from scratch.
3. I could also call `draw_line(...)` from GraphEdit when I click and drag from the area I want. However, to match line's style with the connecting line produced from GraphEdit, I have to code it all over from scratch, as GraphEdit's `_draw_cos_line(...)` is not exposed to gdscript.

### Is there a reason why this should be core and not an add-on in the asset library?

Because I can't find any methods in GraphEdit or GraphNode that lets me drag a port without clicking on it.

### PR DESCRIPTION
**Pull request by mechPenSketch**
Continuation of [this PR](https://github.com/godotengine/godot/pull/51831).

Latest test project: 
[graphedit_hotzones.zip](https://github.com/godotengine/godot/files/7019663/graphedit_hotzones.zip)

Note to self: Always resolve conflicts on Github Desktop, not this site.

_Closes https://github.com/godotengine/godot-proposals/issues/3009_

### CODE REVIEW

**Comment by Gallilus** [comment]
@mechPenSketch 
Made this one for you I remember the first few I did were really painful so I hope this helps you to rebase this PR.
https://youtu.be/iTl8eRAEaqg

**Comment by mechPenSketch** [comment]
OK it looks like the number of files changes is reduced back to the relevant three, following @Gallilus' step. However, I've encountered an error while pulling `godot.windows.tools.64.exe`.

My GPU doesn't seem to be powerful enough for Vulcan

@Gallilus, will this affect testing?

**Comment by Gallilus** [comment]
I gave it a try and checked out the code.
Did not check out core functionality as the basic demo project is not working as intended.

So I gave the code some more debugging so it's easier to see what's going on.

The zoom is not accurately working!?

May I suggest only exposing
`bool GraphEdit::is_in_output_hotzone(GraphNode *p_graph_node, int p_slot_index, const Vector2 &p_mouse_pos, const Vector2i &p_port_size)`.

- left and right are more generic.
- It only requires you to answer the question. Is the mouse close enough to the port to connect?

image
[graphedit_hotzones (4).zip](https://github.com/godotengine/godot/files/7078858/graphedit_hotzones.4.zip)

**Comment by mechPenSketch** [comment]
^Okay. For now, I have commented out `add_virtual_method(get_class_static(), MethodInfo(Variant::BOOL, "_is_in_input_hotzone", ...))`. Hopefully, it's the only change I need to do your further testing.

File: scene/gui/graph_edit.cpp [hunk 1]
```diff
@@ -748,25 +746,29 @@ bool GraphEdit::_check_clickable_control(Control *p_control, const Vector2 &pos)
 	}
 }
 
-bool GraphEdit::is_in_hot_zone(const Vector2 &pos, const Vector2 &p_mouse_pos, const Vector2i &p_port_size, bool p_left) {
```
Comment (Gallilus) [inline scene/gui/graph_edit.cpp:751]: Why not keep this function and make this overwritable?
Instead of creating 2 new ones and overwriting those?

**Comment by Gallilus** [comment]
> ^Okay. For now, I have commented out `add_virtual_method(get_class_static(), MethodInfo(Variant::BOOL, "_is_in_input_hotzone", ...))`. Hopefully, it's the only change I need to do your further testing.

I think there is some communication error. ;)
I mean use is_in_output_hotzone instead of the other to seeing you now have issues with zooming and panning. And will need to change some code anyway.

The last set of ports is not being called.

PS: Took some liberty's with the test project so check that out, maybe double-check if I'm not making any errors.

[graphedit_hotzones (5).zip](https://github.com/godotengine/godot/files/7085490/graphedit_hotzones.5.zip)

https://user-images.githubusercontent.com/20573784/131546425-ca4a98df-4b55-4785-a1d1-a5e06c88fcb9.mp4

File: scene/gui/graph_edit.cpp (hunk 1 above)
Comment (mechPenSketch) [reply scene/gui/graph_edit.cpp:751]: This is to account for whether the user wants to override only one side of the ports.

**Comment by mechPenSketch** [comment]
^ I've looked through your code in your (5) attachment and saw one error (on both functions) that could explain the zooming/panning issue.

On lines 24 and 45,  `node.get_position_offset() + slot.get_position()` should be bracketed together or `zoom` would only be applied to the slot's position.

So on those line, it should be:
`var slot_pos = (node.get_position_offset() + slot.get_position()) * zoom`

**Comment by Gallilus** [comment]
Sorry, meshPenSketch,

I still do not get it to work 
I use `var pan := node.get_position() - node.get_position_offset()`  to compensate for panning but is till need to compensate for the original offset position and `if new_rect.has_point(mouse_position * zoom)` needs to be used as well.

Let me know when you have a functioning project. Then al check on it again

**Comment by Gallilus** [comment]
https://github.com/godotengine/godot-proposals/issues/2790
https://github.com/godotengine/godot-proposals/issues/2886

**Comment by mechPenSketch** [comment]
Maybe other members who are also working on GraphEdit could help.

@Jummit, @Chaosus and @theoway, could you look at @Gallilus' [latest test project](https://github.com/godotengine/godot/files/7085490/graphedit_hotzones.5.zip) to see whether the coordinates are right?

----

Alternatively, if missing test project is preventing this PR from being merged, can I temporary rebase this on the current version (v.3.3 or 3.x)? I can't run the master version as my gpu can't handle vulcan; at best, I can code in blind and hope someone else can test my project to see whether it works.

I'm hoping that by rebasing it on 3.3 or 3.x, I can personally test and make sure I get the co-ordinates right.

**Comment by Gallilus** [comment]
FYI
image

**Comment by Gallilus** [comment]
@mechPenSketch Ohh, I did never replied to your question. Sorry, Man.

Of course, you are able to rebase, it is even preferred.

**Comment by fire** [comment]
Is anything left pending on exposing connection hot zones in GraphNode?

**Comment by mechPenSketch** [comment]
^ Just a test project to see whether it works. I'm rebasing this to `upstream/3.x` to try and get the right co-ordinates by myself.

**Comment by mechPenSketch** [comment]
So, even after rebasing to `upstream/3.x`, I still cannot open godot without vulcan support.

On an unrelated note, I've cleaned up my github desktop from unused branches, but doing so to one of them somehow caused this pr to close. I've managed to resolve it, but it now fails some of the checks.

That being said, I've eventually figured out that I could replicate part of @Gallilus' code on Godot's current version. Specificly, I can _draw() Rect2 that would've represent the hot zones @Gallilus want.

So with that revelation, I can do trial and error to get the right formula.

@Gallilus, can you update your test project with these:

This is the rectangle to be defined for the in hotzone:
```
var slot_pos_in = (graph_node.get_offset() + slot.get_position() - (slot.get_size() * Vector2(0.25, 0))) * zoom - scroll_offset
var rect_size = slot.get_size() * zoom * Vector2(0.5, 1)
var rect_in = Rect2(slot_pos_in, rect_size)
```

And this is the rectangle to be defined for the out hotzone:
```
var slot_pos_out = (graph_node.get_offset() + slot.get_position() + (slot.get_size() * Vector2(0.75, 0))) * zoom - scroll_offset
var rect_size = slot.get_size() * zoom * Vector2(0.5, 1)
var rect_out = Rect2(slot_pos_out, rect_size)
```
For a reference, you can look at this project:
[graphedit_hotzones.zip](https://github.com/godotengine/godot/files/7690251/graphedit_hotzones.zip)

**Comment by mechPenSketch** [comment]
The source of the failed checks seem to be from a wrongly-type constant from an unrelated file:

```
editor/plugins/material_editor_plugin.cpp: In constructor 'MaterialEditor::MaterialEditor()':
editor/plugins/material_editor_plugin.cpp:128:41: error: 'ALIGN_CENTER' is not a member of 'BoxContainer'
  128 |  layout_2d->set_alignment(BoxContainer::ALIGN_CENTER);
      |                                         ^~~~~~~~~~~~
scons: *** [editor/plugins/material_editor_plugin.linuxbsd.tools.64s.o] Error 1
scons: building terminated because of errors.
[Time elapsed: 00:44:54.339]
Error: Process completed with exit code 2.
```
Should I try to correct the code, or is it out of my area? Or is there something else I could do?

**Comment by akien-mga** [comment]
You should rebase, this bug was introduced and fixed 24 h ago, you should happened to tebase at the wrong time.

**Comment by pycbouh** [comment]
Though you probably shouldn't rebase a master PR against `upstream/3.x`...

File: doc/classes/GraphEdit.xml [hunk 2]
```diff
@@ -18,6 +18,25 @@
...
 			</description>
 		</method>
+		<method name="_is_in_input_hotzone" qualifiers="virtual">
+			<return type="bool" />
+			<argument index="0" name="graph_node" type="NodePath" />
+			<argument index="1" name="slot_index" type="int" />
+			<argument index="2" name="mouse_position" type="Vector2" />
+			<description>
+				Returns whether the [code]mouse_position[/code] is in the input hot zone.
+				By default, a hot zone is a [Rect2] positioned such that its center is at [code]get_node(graph_node)[/code].[method GraphNode.get_connection_input_position]([code]slot_index[/code]) (For output's case, call [method GraphNode.get_connection_output_position] instead). The hot zone's width is twice the Theme Property [code]port_grab_distance_horizontal[/code], and its height is twice the [code]port_grab_distance_vertical[/code].
```
Comment (Gallilus) [inline doc/classes/GraphEdit.xml:28]: Mention code to get people started

```
func _is_in_input_hotzone(graph_node, slot_index, mouse_position):
	var slot_size : Vector2 = Vector2(get_theme_constant("port_grab_distance_horizontal"), get_theme_constant("port_grab_distance_vertical"))
	var slot_pos : Vector2 = graph_node.get_position() + graph_node.get_connection_input_position(slot_index) - slot_size / 2
	var rect = Rect2(slot_pos, slot_size)
	
	return rect.has_point(mouse_position)
```

File: doc/classes/GraphEdit.xml [hunk 3]
```diff
@@ -18,6 +18,25 @@
...
+			</description>
+		</method>
+		<method name="_is_in_output_hotzone" qualifiers="virtual">
+			<return type="bool" />
+			<argument index="0" name="graph_node" type="NodePath" />
+			<argument index="1" name="slot_index" type="int" />
+			<argument index="2" name="mouse_position" type="Vector2" />
+			<description>
+				Returns whether the [code]mouse_position[/code] is in the output hot zone. For more information on hot zones, see [method _is_in_input_hotzone].
```
Comment (Gallilus) [inline doc/classes/GraphEdit.xml:37]: Mention code to get people started

func _is_in_output_hotzone(graph_node, slot_index, mouse_position):
	var slot_size : Vector2 = Vector2(get_theme_constant("port_grab_distance_horizontal"), get_theme_constant("port_grab_distance_vertical"))
	var slot_pos : Vector2 = graph_node.get_position() + graph_node.get_connection_output_position(slot_index) - slot_size / 2
	var rect = Rect2(slot_pos, slot_size)
	
	return rect.has_point(mouse_position)

File: scene/gui/graph_edit.cpp [hunk 4]
```diff
@@ -2188,6 +2190,8 @@ void GraphEdit::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("is_right_disconnects_enabled"), &GraphEdit::is_right_disconnects_enabled);
 
 	ClassDB::bind_method(D_METHOD("_update_scroll_offset"), &GraphEdit::_update_scroll_offset);
+	ClassDB::add_virtual_method(get_class_static(), MethodInfo(Variant::BOOL, "_is_in_input_hotzone", PropertyInfo(Variant::NODE_PATH, "graph_node"), PropertyInfo(Variant::INT, "slot_index"), PropertyInfo(Variant::VECTOR2, "mouse_position")));
```
Comment (Gallilus) [inline scene/gui/graph_edit.cpp:2193]: Please pass PropertyInfo(&GraphNode, "graph_node") directly?

File: scene/gui/graph_edit.cpp [hunk 5]
```diff
@@ -2188,6 +2190,8 @@ void GraphEdit::_bind_methods() {
...
+	ClassDB::add_virtual_method(get_class_static(), MethodInfo(Variant::BOOL, "_is_in_output_hotzone", PropertyInfo(Variant::NODE_PATH, "graph_node"), PropertyInfo(Variant::INT, "slot_index"), PropertyInfo(Variant::VECTOR2, "mouse_position")));
```
Comment (Gallilus) [inline scene/gui/graph_edit.cpp:2194]: Please pass PropertyInfo(&GraphNode, "graph_node") directly?

File: doc/classes/GraphEdit.xml (hunk 2 above)
Comment (Gallilus) [inline doc/classes/GraphEdit.xml:28]: Mention code to get people started

```
func _is_in_input_hotzone(graph_node, slot_index, mouse_position):
	var slot_size : Vector2 = Vector2(get_theme_constant("port_grab_distance_horizontal"), get_theme_constant("port_grab_distance_vertical"))
	var slot_pos : Vector2 = graph_node.get_position() + graph_node.get_connection_input_position(slot_index) - slot_size / 2
	var rect = Rect2(slot_pos, slot_size)
	
	return rect.has_point(mouse_position)
```

File: doc/classes/GraphEdit.xml (hunk 3 above)
Comment (Gallilus) [inline doc/classes/GraphEdit.xml:37]: Mention code to get people started

```
func _is_in_output_hotzone(graph_node, slot_index, mouse_position):
	var slot_size : Vector2 = Vector2(get_theme_constant("port_grab_distance_horizontal"), get_theme_constant("port_grab_distance_vertical"))
	var slot_pos : Vector2 = graph_node.get_position() + graph_node.get_connection_output_position(slot_index) - slot_size / 2
	var rect = Rect2(slot_pos, slot_size)
	
	return rect.has_point(mouse_position)
```

**Review by Gallilus — CHANGES_REQUESTED** [review]

**Comment by Gallilus** [comment]
@mechPenSketch Here is my debug scene and code
[Debug_Scene_Graph_Edit_Hot_zone.zip](https://github.com/godotengine/godot/files/7695077/Debug_Scene_Graph_Edit_Hot_zone.zip)

File: scene/gui/graph_edit.cpp (hunk 4 above)
Comment (mechPenSketch) [reply scene/gui/graph_edit.cpp]: Okay. I've updated `PropertyInfo(Variant::NODE_PATH, "graph_node")` accordingly. I've also updated the virtual method call accordingly.

Comment (mechPenSketch) [reply scene/gui/graph_edit.cpp]: I've got 11 failed checks, many point to GraphNode in `PropertyInfo(&GraphNode, "graph_node")` being an illegal use of this type as an expression.

Could this be resolved through `graph_node.h`?
```
./scene/gui/graph_node.h:37:7: note: declared here
class GraphNode : public Container {
      ^
scene/gui/graph_edit.cpp:2220:115: error: 'GraphNode' does not refer to a value
        ClassDB::add_virtual_method(get_class_static(), MethodInfo(Variant::BOOL, "_is_in_output_hotzone", PropertyInfo(&GraphNode, "graph_node"), PropertyInfo(Variant::INT, "slot_index"), PropertyInfo(Variant::VECTOR2, "mouse_position")));
```

Comment (Gallilus) [reply scene/gui/graph_edit.cpp]: Maybe try `Variant::OBJECT`

**Comment by fire** [comment]
@Gallilus I'm not able to review this pr as much as I can, so when it's ready feel free to approve.

File: scene/gui/graph_edit.cpp (hunk 4 above)
Comment (mechPenSketch) [reply scene/gui/graph_edit.cpp]: I've replaced `PropertyInfo(&GraphNode, "graph_node")` with `PropertyInfo(Variant::OBJECT, "graph_node")`.

**Comment by Gallilus** [comment]
@fire  Looks good to me. Thanks @mechPenSketch

**Review by fire — APPROVED** [review]
Looks good to me.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**